javalib/src/main/scala/java/io/BufferedInputStream.scala

```diff
@@ -1,129 +1,247 @@
+  // from spec: "closing a previously closed stream has no effect"
+  // inferred from bug reports: close must be callable async from read
+  override def close(): Unit = Option(in) foreach { source =>
+    // from spec "releases any system resources associated".
+    // implies
+    buf = null
+    source.close()
+    in = null
   }
 
-  override def mark(readLimit: Int): Unit = {
-    if (!closed) {
-      val srcBuf = buf
-      if (buf.size < readLimit)
-        buf = new Array[Byte](readLimit)
+  override def mark(readLimit: Int): Unit = synchronized {
+    marklimit = readLimit
+    markpos = pos
+  }
 
-      // Move data to beginning of buffer
-      if (pos != 0 || (buf ne srcBuf))
-        System.arraycopy(srcBuf, pos, buf, 0, end - pos)
+  override def markSupported(): Boolean = true
 
-      // Update internal state
-      end -= pos
+  /* Reads up to sourceBuffer.length bytes from the source input stream.
+   * This will also replace `buf` if a larger buffer is required to handle mark.
+   *
+   * postcondition: markpos invalidated if pos - markpos exceeds marklimit. Not exactly at the
+   * boundary tho as this is only a "may" in spec.
+   *
+   * @returns buf
+   */
+  private def fillBuffer(sourceBuffer: Array[Byte],
+                         source: InputStream): Option[Array[Byte]] =
+    if (markpos != -1 && (pos - markpos <= marklimit))
+      fillMarkedBuffer(sourceBuffer, source)
+    else
+      fillUnmarkedBuffer(sourceBuffer, source)
+
+  private def fillUnmarkedBuffer(sourceBuffer: Array[Byte],
+                                 source: InputStream): Option[Array[Byte]] = {
+    // mark is always invalidated in this case
+    marklimit = 0
+    markpos = -1
+
+    val bytesRead = source.read(sourceBuffer)
+
+    if (bytesRead == -1) {
+      pos = 0
+      count = 0
+      None
+    } else {
       pos = 0
-      markpos = 0
+      count = bytesRead
+      Option(sourceBuffer)
     }
   }
 
-  override def markSupported(): Boolean = true
-
-  override def read(): Int = {
-    ensureOpen()
+  /* For mark (markpos == -1) the logic is:
+   * If there is space in the current buffer: read into buffer starting at `count`
+   *
+   * If there is no space in the current buffer: create a larger buffer, copy old, read into larger
+   * buffer starting at `count`.
+   *
+   * The mark is not invalidated in this method: Per spec there is no requirement to invalidate mark
+   * *exactly* when pos - markpos exceeds marklimit. This is "generous" and migh permit a reset
+   * beyond marklimit.
+   */
+  private def fillMarkedBuffer(sourceBuffer: Array[Byte],
+                               source: InputStream): Option[Array[Byte]] =
+    if (count < sourceBuffer.length) {
+      fillMarkedBufferUnsafe(sourceBuffer, source)
+    } else {
+      val newBuffer = new Array[Byte](sourceBuffer.length * 2)
+      sourceBuffer.copyToArray(newBuffer)
+      buf = newBuffer
+      fillMarkedBufferUnsafe(newBuffer, source)
+    }
```

Comment:
Consider the simpler:
```scala
val bigEnoughBuf = if (count < sourceBuffer.length) {
  sourceBuffer
} else {
  val newBuffer = new Array[Byte](sourceBuffer.length * 2)
  sourceBuffer.copyToArray(newBuffer)
  buf = newBuffer
  newBuffer
}
// then the current content of fillMarkBufferUnsafe, replacing sourceBuffer by bigEnoughBuf
```

Reply:
Also after setting `buf = newBuffer`, we should check again whether `closed` is true, and if yes, reset it back to `null`:
```scala
buf = newBuffer
if (closed)
  buf = null
```
Doing the naively simpler
```scala
if (!closed)
  buf = newBuffer
```
is not correct, since we could first read `closed` as false, then `close()` would execute and set `closed = true` and `buf = null`, then we'd come back here and do `buf = newBuffer`.